web/src/search_pill.ts

```diff
@@ -112,6 +114,73 @@ function on_pill_exit(
     $user_pill.remove();
 }
 
+// TODO: We're calculating `description_html` every time, even though
+// we only show it (in `generate_pills_html`) for lines with only one
+// pill. We can probably simplify things by separating out a function
+// that generates `description_html` from the information in a single
+// search pill, and remove `description_html` from the `Suggestion` type.
+export function generate_pills_html(suggestion: Suggestion): string {
+    const search_terms = Filter.parse(suggestion.search_string);
+
+    const pill_render_data = search_terms.map((term, index) => {
+        if (user_pill_operators.has(term.operator) && term.operand !== "") {
+            return search_user_pill_data_from_term(term);
+        }
+        const search_pill: SearchPill = {
+            type: "generic_operator",
+            operator: term.operator,
+            operand: term.operand,
+            negated: term.negated,
+        };
+
+        if (search_pill.operator === "topic" && search_pill.operand === "") {
+            return {
+                ...search_pill,
+                is_empty_string_topic: true,
+                sign: search_pill.negated ? "-" : "",
+                topic_display_name: util.get_final_topic_display_name(""),
+            };
+        }
+        if (search_pill.operator === "search") {
+            let description_html = search_term_description_html(search_pill);
+            // We capitalize the beginning of the suggestion line if it's text (not
+            // pills), which is only relevant for suggestions with search operators.
+            if (index === 0) {
+                const capitalized_first_letter = description_html.charAt(0).toUpperCase();
+                description_html = capitalized_first_letter + description_html.slice(1);
+            }
```

Comment:
This algorithm can't possibly be correct in the general case: If the HTML starts with `<`, it will capitalize the wrong thing.

It's probably OK to merge though. I think we should just do the follow-up of redoing the descriptions to all start with a capital letter and be internationalized; now that we're no longer combining/composing them, we should be able to do normal internationalization (well, with just two strings for each operator, one for negated and one for not).

Reply:
I was copying existing logic in `finalize_search_result`, but agreed the general pattern could be set up better.